### PR TITLE
Added docker-service-kill experiment implementation in litmus-go.

### PR DESCRIPTION
**Pull request by Jonsy13**
Signed-off-by: Jonsy13 <vedant.shrotria@chaosnative.com>

**What this PR does / why we need it**:

**This PR will add the implementation of `docker-service-kill` in litmus-go.**

**Some observations while testing on GKE Cluster -** 
 - While working with Image type as `Container-Optimised OS with Containerd`, It is only killing the docker service, but Node is still in Ready state.
 - While working with Image type as `Ubuntu with Docker (Ubuntu)`, It is killing the docker service as well as Node goes in NotReady State.

**Checklist:**
-   [ ] Fixes #<issue number>
-   [ ] PR messages has document related information
-   [ ] Labelled this PR & related issue with `breaking-changes` tag
-   [ ] PR messages has breaking changes related information
-   [ ] Labelled this PR & related issue with `requires-upgrade` tag
-   [ ] PR messages has upgrade related information
-   [ ] Commit has unit tests
-   [ ] Commit has integration tests
-   [ ] E2E run Required for the changes

Experiment Logs - (Ubuntu with Docker (Ubuntu))
```
➜  experiment git:(dockersvc) ✗ go run experiment.go --name "docker-service-kill" --kubeconfig ~/.kube/config
INFO[2021-06-11T11:49:12+05:30] Experiment Name: docker-service-kill         
INFO[2021-06-11T11:49:12+05:30] [PreReq]: Getting the ENV for the  experiment 
INFO[2021-06-11T11:49:12+05:30] [PreReq]: Updating the chaos result of docker-service-kill experiment (SOT) 
INFO[2021-06-11T11:49:14+05:30] [Info]: The application information is as follows  Node Label= Target Node= Ramp Time=0 Namespace= App Label=
INFO[2021-06-11T11:49:14+05:30] [Status]: Verify that the AUT (Application Under Test) is running (pre-chaos) 
INFO[2021-06-11T11:49:14+05:30] [Status]: No appLabels provided, skipping the application status checks 
INFO[2021-06-11T11:49:14+05:30] [Status]: Getting the status of target nodes 
INFO[2021-06-11T11:49:15+05:30] The Node status are as follows                Ready=true Node=gke-vedant-2-default-pool-5ba8b8ef-8mjh
INFO[2021-06-11T11:49:15+05:30] The Node status are as follows                Node=gke-vedant-2-default-pool-5ba8b8ef-cfsq Ready=true
INFO[2021-06-11T11:49:15+05:30] The Node status are as follows                Ready=true Node=gke-vedant-2-default-pool-5ba8b8ef-fv9q
INFO[2021-06-11T11:49:15+05:30] [Info]: Details of node under chaos injection  NodeName=gke-vedant-2-default-pool-5ba8b8ef-cfsq
INFO[2021-06-11T11:49:16+05:30] [Status]: Checking the status of the helper pod 
INFO[2021-06-11T11:49:18+05:30] docker-service-kill-helper-xtegvx helper pod is in Running state 
INFO[2021-06-11T11:49:20+05:30] [Status]: Check for the node to be in NotReady state 
INFO[2021-06-11T11:49:33+05:30] The Node status are as follows                Node=gke-vedant-2-default-pool-5ba8b8ef-cfsq Ready=false
INFO[2021-06-11T11:49:33+05:30] [Wait]: Waiting till the completion of the helper pod 
INFO[2021-06-11T11:49:34+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:35+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:37+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:38+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:40+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:41+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:43+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:44+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:45+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:47+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:48+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:49+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:51+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:52+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:53+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:55+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:56+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:58+05:30] helper pod status: Running                   
INFO[2021-06-11T11:49:59+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:00+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:02+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:03+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:04+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:06+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:07+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:08+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:10+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:11+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:13+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:14+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:15+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:17+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:18+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:20+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:21+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:23+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:24+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:26+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:27+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:28+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:30+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:31+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:33+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:34+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:36+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:37+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:38+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:40+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:41+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:43+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:44+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:46+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:47+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:49+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:50+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:52+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:53+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:54+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:56+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:57+05:30] helper pod status: Running                   
INFO[2021-06-11T11:50:59+05:30] helper pod status: Running                   
INFO[2021-06-11T11:51:00+05:30] helper pod status: Running                   
INFO[2021-06-11T11:51:01+05:30] helper pod status: Succeeded                 
INFO[2021-06-11T11:51:01+05:30] [Status]: The running status of Pods are as follows  Pod=docker-service-kill-helper-xtegvx Status=Succeeded
INFO[2021-06-11T11:51:02+05:30] [Status]: Getting the status of target nodes 
INFO[2021-06-11T11:51:03+05:30] The Node status are as follows                Ready=true Node=gke-vedant-2-default-pool-5ba8b8ef-cfsq
INFO[2021-06-11T11:51:03+05:30] [Cleanup]: Deleting the helper pod           
INFO[2021-06-11T11:51:05+05:30] [Confirmation]: docker-service-kill chaos has been injected successfully 
INFO[2021-06-11T11:51:05+05:30] [Status]: Verify that the AUT (Application Under Test) is running (post-chaos) 
INFO[2021-06-11T11:51:05+05:30] [Status]: No appLabels provided, skipping the application status checks 
INFO[2021-06-11T11:51:05+05:30] [The End]: Updating the chaos result of docker-service-kill experiment (EOT) 

```

### CODE REVIEW

File: chaoslib/litmus/docker-service-kill/lib/docker-service-kill.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	"github.com/litmuschaos/litmus-go/pkg/types"
+	"github.com/litmuschaos/litmus-go/pkg/utils/common"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PrepareDockerKill contains prepration steps before chaos injection
+func PrepareDockerKill(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
```
Comment (ispeakc0de) [inline chaoslib/litmus/docker-service-kill/lib/docker-service-kill.go:21]: PrepareDockerKill --> PrepareDockerServiceKill

File: chaoslib/litmus/docker-service-kill/lib/docker-service-kill.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+		events.GenerateEvents(eventsDetails, clients, chaosDetails, "ChaosEngine")
+	}
+
+	if experimentsDetails.EngineName != "" {
+		if err := setHelperData(experimentsDetails, clients); err != nil {
+			return err
+		}
+	}
+
+	// Creating the helper pod to perform node memory hog
```
Comment (ispeakc0de) [inline chaoslib/litmus/docker-service-kill/lib/docker-service-kill.go:56]:
```suggestion
	// Creating the helper pod to perform docker-service-kill
```

File: experiments/generic/docker-service-kill/test/test.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          # provide application kind
+          - name: APP_KIND
+            value: '' 
+
+          - name: TOTAL_CHAOS_DURATION
+            value: ''
+
+          # provide auxiliary application details - namespace and labels of the applications
+          # sample input is - "ns1:app=percona,ns2:name=nginx"
+          - name: AUXILIARY_APPINFO
```
Comment (ispeakc0de) [inline experiments/generic/docker-service-kill/test/test.yml:41]: We can add the other envs i.e NODE_LABEL, TARGET_NODE

**Review by ispeakc0de — DISMISSED** [review]

**Review by ispeakc0de — DISMISSED** [review]

**Comment by uditgaurav** [comment]
We can also try to add the support for `Containerd ubuntu` image. In that case, we have to wait for the docker service to be up and active also skip the node status check (as node will not get down). We can do this in upcoming PR also. -- @Jonsy13

File: experiments/generic/docker-service-kill/README.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+<table>
+<tr>
+<th> Name </th>
+<th> Description </th>
+<th> Documentation Link </th>
+</tr>
+<tr>
+ <td> Docker Service Kill </td>
+ <td> This experiment causes docker service kill gracefully for a certain chaos duration. The experiment aims to verify resiliency of applications whose replicas may be evicted or becomes unreachable on account on nodes turning unschedulable (Not Ready) due to docker service kill. </td>
+ <td>  <a href=""> Added soon </a> </td>
```
Comment (uditgaurav) [inline experiments/generic/docker-service-kill/README.md:12]:
```suggestion
 <td>  <a href="https://docs.litmuschaos.io/docs/docker-service-kill/"> Here </a> </td>
```

File: experiments/generic/docker-service-kill/experiment/docker-service-kill.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+	result.SetResultUID(&resultDetails, clients, &chaosDetails)
+
+	// generating the event in chaosresult to marked the verdict as awaited
+	msg := "experiment: " + experimentsDetails.ExperimentName + ", Result: Awaited"
+	types.SetResultEventAttributes(&eventsDetails, types.AwaitedVerdict, msg, "Normal", &resultDetails)
+	events.GenerateEvents(&eventsDetails, clients, &chaosDetails, "ChaosResult")
+
+	//DISPLAY THE APP INFORMATION
+	log.InfoWithValues("[Info]: The application information is as follows", logrus.Fields{
+		"Namespace":   experimentsDetails.AppNS,
```
Comment (uditgaurav) [inline experiments/generic/docker-service-kill/experiment/docker-service-kill.go:64]:
```suggestion
		"App Namespace":   experimentsDetails.AppNS,
```

File: experiments/generic/docker-service-kill/experiment/docker-service-kill.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+			failStep := "failed in chaos injection phase"
+			result.RecordAfterFailure(&chaosDetails, &resultDetails, failStep, clients, &eventsDetails)
+			log.Errorf("Chaos injection failed, err: %v", err)
+			return
+		}
+	default:
+		failStep := "lib and container-runtime combination not supported!"
+		result.RecordAfterFailure(&chaosDetails, &resultDetails, failStep, clients, &eventsDetails)
+		log.Error("lib and container-runtime combination not supported, provide the correct value of lib & container-runtime")
+		return
```
Comment (uditgaurav) [inline experiments/generic/docker-service-kill/experiment/docker-service-kill.go:139]:
```suggestion
				log.Error("[Invalid]: Please Provide the correct LIB")
		failStep := "no match found for specified lib"
		result.RecordAfterFailure(&chaosDetails, &resultDetails, failStep, clients, &eventsDetails)
		return
```
We are not using container runtime parameter for docker svc kill.

**Review by uditgaurav — APPROVED** [review]

**Review by ispeakc0de — APPROVED** [review]

**Review by ksatchit — APPROVED** [review]

**Pull request merged**